src/main/java/org/broadinstitute/hellbender/utils/bwa/BwaMemIndex.java

```diff
@@ -30,12 +32,55 @@
  * if that's what you want to do.
  */
 public final class BwaMemIndex implements AutoCloseable {
+
+    /**
+     * Indexing algorithms supported by Bwa.
+     */
+    public enum Algorithm {
+        RB2, IS, AUTO;
+
+        /**
+         * The string name use by the Bwa command line to denote that algorithm.
+         * @return never {@code null} and unique across algorithms.
+         */
+        public String toBwaName() {
+            return this.toString().toLowerCase();
+        }
+    }
+
     private final String indexImageFile; // stash this for error messages
     private volatile long indexAddress; // address where the index was memory-mapped (for use by C code)
     private final AtomicInteger refCount; // keep track of how many threads are actively aligning
     private final List<String> refContigNames; // the reference dictionary from the index
     private static volatile boolean nativeLibLoaded = false; // whether we've loaded the native library or not
 
+    private static final List<String> FASTA_EXTENSIONS = Collections.unmodifiableList(Arrays.asList(".fasta", ".fa"));
+
+
+    public static void indexReference(final String referenceName, final String indexPrefix, final Algorithm algo) {
+        final String referenceFileName = resolveReferenceFileName(referenceName);
+        loadNativeLibrary();
+        createReferenceIndex(referenceFileName, indexPrefix, algo.toBwaName());
+    }
+
+    private static String resolveReferenceFileName(final String referenceName) {
+        if (referenceName == null) {
+            throw new IllegalArgumentException("the reference name provided cannot be null");
+        }
+        final String referenceFileName = FASTA_EXTENSIONS.stream()
+                .filter(ext -> new File(referenceName + ext).isFile()).findFirst().orElse(referenceName);
+        final File referenceFile = new File(referenceFileName);
+        if (referenceFile.isFile()) {
+            if (referenceFile.canRead()) {
+                return referenceFile.getPath();
+            } else {
+                throw new IllegalArgumentException("the reference file provided is not readable: " + referenceFile);
```

Comment:
I've often wanted to push the idea of UserExceptions into the rest our dependencies.  Maybe we should make an extremely minimal UserException artifact that all of our little related projects can depend on...  It neatly divides the "things that are the users fault" from the "things that are the developers fault"